torch/utils/checkpoint.py

```diff
@@ -37,17 +57,32 @@ def backward(ctx, *args):
         if not torch.autograd._is_checkpoint_valid():
             raise RuntimeError("Checkpointing is not compatible with .grad(), please use .backward() if possible")
         inputs = ctx.saved_tensors
+        if ctx.preserve_rng_state:
+            # Stash the surrounding rng state, and mimic the state that was
+            # present at this time during forward.  If cuda was not initialized
+            # at this time during the original forward, we assume that the imminent
+            # forward-segment-within-backward will not alter the cuda rng state.
+            current_cpu_rng_state = torch.get_rng_state()
+            torch.set_rng_state(ctx.fwd_cpu_rng_state)
+            if ctx.had_cuda_in_fwd:
+                current_cuda_rng_state = torch.cuda.get_rng_state()
+                torch.cuda.set_rng_state(ctx.fwd_cuda_rng_state)
```

Comment:
Can we implement a helper like `get_rng_states()` and `set_rng_states()` that deals with CUDA and other things internally? It could be a regular Python class that encapsulates all the logic. Also, it would be good to use context managers, so that we do restore the correct thing when an exception is thrown.

Reply:
Do you want me to do that as part of this PR?

Reply:
Not sure how well this would fit but may be you could use `with torch.random.fork_rng():`? This exists in `torch/random.py`

Reply:
@syed-ahmed Good find, this looks perfect.